# Introduction

In this project, we will work with a dataset of used cars from eBay Kleinanzeigen, a classifieds sections of the German eBay website.

The data dictionary provided with data is as follows:

- __dateCrawled__ - When this ad was first crawled. All field-values are taken from this date.
-  __name__ - Name of the car.
-  __seller__ - Whether the seller is private or a dealer.
-  __offerType__ - The type of listing
-  __price__ - The price on the ad to sell the car.
-  __abtest__ - Whether the listing is included in an A/B test.
-  __vehicleType__ - The vehicle Type.
-  __yearOfRegistration__ - The year in which the car was first - - - registered.
-  __gearbox__ - The transmission type.
-  __powerPS__ - The power of the car in PS.
-  __model__ - The car model name.
-  __odometer__ - How many kilometers the car has driven.
-  __monthOfRegistration__ - The month in which the car was first registered.
-  __fuelType__ - What type of fuel the car uses.
-  __brand__ - The brand of the car.
-  __notRepairedDamage__ - If the car has a damage which is not yet repaired.
-  __dateCreated__ - The date on which the eBay listing was created.
-  __nrOfPictures__ - The number of pictures in the ad.
-  __postalCode__ - The postal code for the location of the vehicle.
-  __lastSeenOnline__ - When the crawler saw this ad last online.


The aim here is to clean and analyze the data. 

Let's get to it.

In [1]:
#Import libraries that we will be using
import pandas as pd
import numpy as np

autos = pd.read_csv("datasets/autos.csv", encoding="Latin-1")

In [2]:
#Exploring out datasets
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
#Exploring some info about our dataset
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

For this dataset we have 20 columns, most of wich are strings, with 50000 observations (rows). We can see that columns __vehicleType__, __gearbox__, __model__, __fuelType__ and __notRepairedDamage__ have some entries with Null values. We need to take care of those values.

# Cleaning Columns Names

We need to convert the columns names from [camelcase](https://en.wikipedia.org/wiki/Camel_case) to [snakecase](https://en.wikipedia.org/wiki/Snake_case) and reword some of the column names based on the data dictionary to be more descriptive.

In [4]:
#The columns names that we have at the moment
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
#Let's convert the columns names
#yearOfRegistration to registration_year
#monthOfRegistration to registration_month
#notRepairedDamage to unrepaired_damage
#dateCreated to ad_created
#The rest of the columnn names from camelcase to snakecase.

cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.columns = cols

In [6]:
#Let's take a look at the new columns names
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We change the columns names to names that are more expressive, more informative. We need to look at the column name and know right away what that column is about.  

# Initial Exploration and Cleaning

Let's do some basic data exploration to determine what other cleaning tasks need to be done. We will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [7]:
#Descriptive statistics for all columns
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

We can see that the __nr_of_pictures__ column has mean equals 0, therefore this columns has no values, we can drop it.

The __price__ and __odometer__ columns are numeric values stored as text. We need to:

- Remove any non-numeric characters from them.
- Convert the column to a numery dtype.
- Rename the column __odometer__ to __odometer_km__.

In [8]:
#Drop nr_of_pictures column
autos = autos.drop("nr_of_pictures", axis=1)

In [9]:
#Adjusting price and odometer columns

#Removing $ character
autos["price"] = autos["price"].str.replace("$","")

#Replacing , (comma) for no-space and casting to int
autos["price"] = autos["price"].str.replace(",", "").astype(int)

In [10]:
#Removing km character
autos["odometer"] = autos["odometer"].str.replace("km","")

#Replacing , (comma) for no-space and casting to int
autos["odometer"] = autos["odometer"].str.replace(",", "").astype(int)

In [11]:
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

In [12]:
autos[:5]

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00        79588   
1         bmw              nein  2016-04-04 00:00:00        71034   
2  volkswagen              nein  2016-03-26 00:00:00        35394   
3       smart              nein  2016-03-12 00:00:00        33729   
4        ford              nein  2016-04-01 00:00:00        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

# Exploring the Odometer and Price Columns

Let's continue exploring the data, specifically looking for data that doesn't look right.

We will start by analyzing the __odometer_km__ and __price__ columns.

We will check for:

- Using minimum and maximum values we're going to look for any values that look unrealistically high or low (outliers) that we might want to remove.

- We will also use:
    - unique().shape to see how many unique values
    - describe() to view min/max/median/mean
    - value_counts(), with some variations.

In [ ]:
#Let's explore odometer_km first
max_value = autos["odometer_km"].max()
min_value = autos["odometer_km"].min()

print("Max Value: ", max_value)
print("Min Value: ", min_value)

print("Shape: ", autos["odometer_km"].shape)
print("Describe: ", autos["odometer_km"].describe())
print("Head: \n", autos["odometer_km"].head(10))
print("Count: ", autos["odometer_km"].value_counts().head(20))

The values for Max and Min of odometer_km look like realistic numbers.

We can see that the most cars have 125000 km driven, it is a lot of cars, 5170.

In [ ]:
#Let's explore price
max_value = autos["price"].max()
min_value = autos["price"].min()

print("Max Value: ", max_value)
print("Min Value: ", min_value)

print("Shape: ", autos["price"].shape)
print("Describe: ", autos["price"].describe())
print("Head: \n", autos["price"].head(10))
print("Count: ", autos["price"].value_counts().head(20))

For the price we can see the $500 cars were the most bought, 781 people bought.

There are 1,421 cars listed with $0 price (free car? I want one) - given that this is only 2% of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [ ]:
#More about prices - Most expensives
autos["price"].value_counts().sort_index(ascending=False).head(20)

In [ ]:
#More cheap
autos["price"].value_counts().sort_index(ascending=True).head(20)

We can see that we have car with a high price. 
We also have cars for \$0 and cars for $1, \$2 ...

Given that eBay is an auction site, there could legitimately be items where the opening bid is $1. \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [ ]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

# Exploring the date Columns

Let's now explore the date columns and understand the date range the data covers.

There are 5 columns that should represent date values:

- __date_crawled__: addeb by the crawler
- __last_seen__: added by the crawler
- __ad_created__: from the website
- __registration_month__: from the website
- __registration_year__: from the website

The columns __date_crawled__, __last_seen__ and __ad_created__ are identified as string values by pandas.

We need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric valuesm so we can use methods like describe() to understand the distribution without any extra data processing.

But first, let's see and understand how the values in the three string columns are formatted.

In [ ]:
#Take a look at the three columns
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

We can notice that the first 10 characters represent the day, like in row 0 (2016-03-26).

We can extract just the date values and try to understand the data range.

In [ ]:
#Extracting the date values
print(autos["date_crawled"].str[:10][:10])

In [ ]:
#Let's calculate the distribution of values in
#date_crawled column
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

For the result we have date started in March 0f 2016 until April of 2016.

In [ ]:
#Let's calculate the distribution of values in
#ad_created
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

For the ad_created we have date started in June of 2016 until March of 2016.

In [ ]:
#Let's calculate the distribution of values in
#last_seen columns
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

For last_seen we have date started in March of 2016 until April of 2016.

In [ ]:
#Take a look at registration_year
autos["registration_year"].describe()

We have some unusual values for registration years, like the max value equals to 9999 (there is no such year registered, not yet) and we also have the year 1000 (I think that back then we did not have cars yet). We need to ged rid of these values.

# Dealing with Incorrect Registration Year Data

We seen that we have incorrect years in our registration_year column, we need to fix them.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

One option is to remove the values that are not part of this interval (1900 - 2016). Let's how many incorrect values we have between those years.

In [ ]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

We have almost 4%, it is a little portion, so we can remove.

In [ ]:
#We can use between to get those values
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

We can see that the most of vehicles were registered in the last 20 years.

# Exploring Price by Brand

Let's verify what brands have the most expensive values.

In [ ]:
#The brands
count_brands = autos['brand'].value_counts()
count_brands

In [ ]:
#Let's do it with brands >= 1000 values
price_by_brand = {}
brands = count_brands[count_brands > 1000].index

for row in brands:    
    selected_rows = autos.loc[autos['brand'] == row, 'price']
    selected_rows = selected_rows.mean()
    price_by_brand[row] = selected_rows

In [ ]:
price_by_brand

In [ ]:
sorted(price_by_brand.items(), key=lambda x: x[1], reverse=True)

We can see that audi, mercedes bez and bmw are the most expensive brans, while open, fiat and renault are the most accessible one.

# Storing Aggregate Data in a DataFrame

In the last screen, we aggregated across brands to understand mean price. We observed that in the top 6 brands, there's a distinct price gap.

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive

- Volkswagen is in between

In [ ]:
#Exploring mileage
bmp_series = pd.Series(price_by_brand)
pd.DataFrame(bmp_series, columns=["mean_mileage"])

In [ ]:
brand_mean_mileage = {}

for brand in brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(price_by_brand).sort_values(ascending=False)

In [ ]:
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info

In [ ]:
brand_info["mean_price"] = mean_prices
brand_info

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.